From 968a04e9006f628b1ca5839cb726837e06372a2a Mon Sep 17 00:00:00 2001
From: Li Wencheng <liwencheng@phytium.com.cn>
Date: Wed, 7 May 2025 17:22:12 +0800
Subject: [PATCH 31/67] PHYTIUM: net/phytmac: Slove left-shift out of bound
 issue

When the value of the bd_prefetch is equal to 0, subtracting 1 and
shifting left will cause the overflow issue.

Signed-off-by: Li Wencheng <liwencheng@phytium.com.cn>
Signed-off-by: Wang Yinfeng <wangyinfeng@phytium.com.cn>
Change-Id: Idc6b8f8100c62331ca6e9778082c533221558e8a

Link: https://gitee.com/phytium_embedded/phytium-linux-kernel/commit/ec5f2c26458fcf71564603c01901a717eeddee19
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 drivers/net/ethernet/phytium/phytmac.h    |  2 +-
 drivers/net/ethernet/phytium/phytmac_v2.c | 11 +++++++----
 2 files changed, 8 insertions(+), 5 deletions(-)

diff --git a/drivers/net/ethernet/phytium/phytmac.h b/drivers/net/ethernet/phytium/phytmac.h
index 3806e09bc122..4d66ae498b88 100644
--- a/drivers/net/ethernet/phytium/phytmac.h
+++ b/drivers/net/ethernet/phytium/phytmac.h
@@ -15,7 +15,7 @@

 #define PHYTMAC_DRV_NAME		"phytium-mac"
 #define PHYTMAC_DRV_DESC		"PHYTIUM Ethernet Driver"
-#define PHYTMAC_DRIVER_VERSION		"1.0.29"
+#define PHYTMAC_DRIVER_VERSION		"1.0.39"
 #define PHYTMAC_DEFAULT_MSG_ENABLE	  \
 		(NETIF_MSG_DRV		| \
 		NETIF_MSG_PROBE	| \
diff --git a/drivers/net/ethernet/phytium/phytmac_v2.c b/drivers/net/ethernet/phytium/phytmac_v2.c
index b8af75b3d934..9bab3d2a5ca4 100644
--- a/drivers/net/ethernet/phytium/phytmac_v2.c
+++ b/drivers/net/ethernet/phytium/phytmac_v2.c
@@ -342,10 +342,13 @@ static int phytmac_v2_get_feature_all(struct phytmac *pdata)
 		pdata->dma_addr_width = 32;
 	pdata->dma_data_width = para.dma_data_width;
 	pdata->max_rx_fs = para.max_rx_fs;
-	pdata->tx_bd_prefetch = (2 << (para.tx_bd_prefetch - 1)) *
-				sizeof(struct phytmac_dma_desc);
-	pdata->rx_bd_prefetch = (2 << (para.rx_bd_prefetch - 1)) *
-				sizeof(struct phytmac_dma_desc);
+
+	if (para.tx_bd_prefetch)
+		pdata->tx_bd_prefetch = (2 << (para.tx_bd_prefetch - 1)) *
+					sizeof(struct phytmac_dma_desc);
+	if (para.rx_bd_prefetch)
+		pdata->rx_bd_prefetch = (2 << (para.rx_bd_prefetch - 1)) *
+					sizeof(struct phytmac_dma_desc);

 	if (netif_msg_hw(pdata)) {
 		netdev_info(pdata->ndev, "feature qnum=%d, daw=%d, dbw=%d, rxfs=%d, rxbd=%d, txbd=%d\n",
--
2.50.1
